secret-store/src/key_server_cluster/cluster_sessions.rs

```diff
@@ -569,6 +616,53 @@ impl ClusterSession for AdminSession {
 	}
 }
 
+impl<S: ClusterSession> WaitableSession<S> {
+	pub fn new(session: S, oneshot: Oneshot<Result<S::SuccessfulResult, Error>>) -> Self {
+		WaitableSession {
+			session: Arc::new(session),
+			oneshot,
+		}
+	}
+
+	pub fn into_wait_future(self) -> Box<Future<Item=S::SuccessfulResult, Error=Error> + Send> {
+		Box::new(self.oneshot
+			.map_err(|e| Error::Internal(e.to_string()))
+			.and_then(|res| res))
+	}
+}
+
+#[cfg(test)]
+impl<T> CompletionSignal<T> {
```

Comment:
Some docs explaining why we use two impls and how they differ? I'd really hate to have some dev down the line change one and leave the other and end up with passing tests and broken impl...

Reply:
…or even a macro generating the right code?

Reply:
I've merged implementations